src/poly/mod.rs

```diff
@@ -14,38 +14,88 @@ pub mod mielim;
 pub mod reduce;
 pub mod simplify;
 
-pub trait ToExpr<F, V, M> {
+pub trait ToExpr<F, V, M: Meta> {
     fn expr(&self) -> Expr<F, V, M>;
 }
 
 pub trait ToField<F> {
     fn field(&self) -> F;
 }
 
+pub trait Meta: Clone + Default + Debug + Hash {
```

Comment:
Cool trick with attaching meta! I have a question: do you think there could be any other use for Meta in the future besides hashing? If not, we could rename it into "HashingStrategy", and rename "from_expr" into something like "compute_hash".

Reply:
Yes, we will use meta to keep DebugSymRefs all the way to the SBPIR, so the MockProver on SBPIR can show good error messages

Reply:
You mean something like
```rust
impl Meta for DebugSymRef {
    fn from_expr<F: Field + Hash, V: Clone + Eq + Hash>(
        expr: &Expr<F, V, ()>,
        assignments: &VarAssignments<F, V>,
    ) -> Self {
        //create a debug sym from expr and assignments
       ...
    }
}
```

Reply:
I'm not sure if I fully understand the purpose but my concern is that it looks like "anything factory" because `from_expr` only tells that "something could be made of an expression and assignments". Also, we have to be sure that we won't need the hash and the debug symbols at the same time as there is only one Meta in an expression.

Reply:
I am not sure, I was going to dive into the details once I am sure about the algorithm itself. What I mean is the M type parameter in expression. We will use it for DebugSymRef and other things, like in this case for the CSE.

You can do mappings that change the content of M.

Reply:
This is what concerns me. It looks overly generic as it can be a different thing at different times.

Reply:
What is the issue of being too generic? If you just plug in some metadata that you need at a time and then when you don't need anymore you just remove it, I think it's ok to be generic. But I can be wrong

Reply:
> I'm not sure if I fully understand the purpose but my concern is that it looks like "anything factory" because `from_expr` only tells that "something could be made of an expression and assignments". Also, we have to be sure that we won't need the hash and the debug symbols at the same time as there is only one Meta in an expression.

but I do agree the name is not the best, I was at first calling `apply_meta` and also it should not depend on the `assignments` for everything, so I need to change that

Reply:
Why would a `M` would need always assignemtns to be created?

I am not sure this is a method that we would need for any `M` on a expression

Reply:
> Why would a `M` would need always assignemtns to be created?
> 
> I am not sure this is a method that we would need for any `M` on a expression

Yeah it's not, I'm gonna change it